### PR TITLE
Refactor of instance conververter codegen

### PR DESCRIPTION
**Pull request by grechaw**
New public methods #301 
Removed verbose comments #297
Refactored modules to use separate let and return #232
need to review and circulate

version translator cleanup, refactor

Style refactor of version translator

This PR contains no functional changes to ES, but prepares for some work on JSON envelopes and JavaScript.  It also incorporates feedback from @paxtonhare and @matt-turner about code style.

This PR is for discussion -- once we've got consensus in the group I'll update the spec and circulate for review.

Here's is a "new style" extraction function:

```XQuery
(:~
 : Extracts instance data, as a map:map, from some source document.
 : @param $source-node  A document or node that contains
 :   data for populating a ETOne
 : @return A map:map instance with extracted data and
 :   metadata about the instance.
 :)
declare function conversion:extract-instance-ETOne(
    $source as node()?
) as map:map
{
    let $source-node := es:init-source($source, 'ETOne')
    (: begin customizations here :)
    let $pk  :=             $source-node/pk ! sem:iri(.)
    let $p1  :=             $source-node/p1 ! xs:integer(.)
    let $p2  :=             $source-node/p2 ! xs:integer(.)
    let $p3  :=             es:extract-array($source-node/p3, xs:integer#1)
    let $p4  :=             es:extract-array($source-node/p4, xs:integer#1)
    let $p5  :=             $source-node/p5 ! xs:integer(.)
    let $p6  :=             $source-node/p6 ! xs:integer(.)
    (: end customizations :)

    let $instance := es:init-instance($source-node, 'ETOne')
    (: Comment or remove the following line to suppress attachments :)
        =>es:add-attachments($source-node, $source)

    return
    if (empty($source-node/*)) 
    then $instance
    else $instance
        =>   map:with('pk', $pk)
        =>   map:with('p1', $p1)
        =>   map:with('p2', $p2)
        =>es:optional('p3', $p3)
        =>es:optional('p4', $p4)
        =>es:optional('p5', $p5)
        =>es:optional('p6', $p6)
};
```

Things to note -- 

* uses new public functions to init source, init instance, and add attchments.
* All XPath expressions are at the top of function, and populate variables.
* Section at the bottom is expected to not have to change -- obvious how the map is constructed.

### CODE REVIEW

**Comment by grechaw** [comment]
Assigned this PR to three of us for visibility and requested a review from @paxtonhare

File: entity-services/src/main/xdmp/entity-services/entity-services-codegen.xqy [hunk 1]
```diff
@@ -197,50 +192,36 @@ declare function {$prefix}:extract-instance-{$entity-type-name}(
     $source as node()?
```
Comment (paxtonhare) [inline entity-services/src/main/xdmp/entity-services/entity-services-codegen.xqy]: I know this was already `$source as node()?` but I wanted to point out that json:object() and json:array() are not node(). They are item(). So this means you can't pass in json to this generated function. For the DHF I changed it to item()? to be more accepting of others.

**Comment by paxtonhare** [comment]
another thing to consider... the code, as generated will have unix line endings. I just updated the DHF codegen to spit out \r\n so that windows users w/o access to decent editors (more common than you care to think) can actually see the code w/o it being one huge, ugly line.

**Comment by paxtonhare** [comment]
please make esi:wrap-duplicates public. I use that in the DHF and have to rewrite it because it is currently private.

File: entity-services/src/main/xdmp/entity-services/entity-services-codegen.xqy [hunk 2]
```diff
@@ -197,50 +192,36 @@ declare function {$prefix}:extract-instance-{$entity-type-name}(
...
-          cts:search( collection('customers'),
-              string($source-node/path/to/lookup/key))/id
-    or populate the instance with constants.
-    =>map:with('constantValue', 10)
-    The output of this function should structurally match the output of
-    es:model-get-test-instances($model)
-    :)
-
-    $instance
+    if (empty($source-node/*))
```
Comment (paxtonhare) [inline entity-services/src/main/xdmp/entity-services/entity-services-codegen.xqy]: why the if? can't you just return $instance

File: entity-services/src/main/xdmp/entity-services/entity-services-codegen.xqy (hunk 1 above)
Comment (grechaw) [inline entity-services/src/main/xdmp/entity-services/entity-services-codegen.xqy]: Makes sense to me, I'll update with that change

**Comment by grechaw** [comment]
OK -- seems I should make wrap-duplicates public, but leave it in the same module without a call in the public API.  It seems very 'helperish' to the implementation.  Do you agree @paxtonhare ?

**Comment by paxtonhare** [comment]
@grechaw yes. I agree. make it public, but don't advertise it in the main module.

File: entity-services/src/main/xdmp/entity-services/entity-services-codegen.xqy (hunk 2 above)
Comment (grechaw) [inline entity-services/src/main/xdmp/entity-services/entity-services-codegen.xqy]: Well, I agree this looks rather inelegant, and I tripped up on it before.

init-instance returns a structure like this if the source is a reference  (a source is assumed to be a reference if there are no children of the type node... ie `<friendOf><Person>112</Person></friendOf>` would be extracted returned from es:init-instance() as 

`{"$type":"Person":"$ref":112}`

So if the return value from init-instance has this "$ref" key, I don't want to process all of the setters.

**Pull request merged**